app/components/file-viewer/canvas-viewer.jsx

```diff
@@ -0,0 +1,111 @@
+import React from 'react';
+import Markdown from 'markdownz';
+import LoadingIndicator from '../loading-indicator';
+import modelSelector from '../modelling';
+import alert from '../../lib/alert';
+
+class CanvasViewer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.onLoad = this.onLoad.bind(this);
+    this.state = {
+      loading: true
+    };
+  }
+  componentDidMount() {
+    // add the canvas and prep for rendering
+    this.createNewModel(this.props);
+  }
+  shouldComponentUpdate(nextProps, nextState) {
+    // check if a new subject has been provided
+    if (this.state.loading !== nextState.loading || this.props.src !== nextProps.src) {
+      return true;
+    }
+    // Only re-render when annotation has changed, or we have zoomed/panned
+    // JSON is expensive, so reduce the test as much as possible
+    if (JSON.stringify(nextProps.annotations) !== JSON.stringify(this.props.annotations)) {
+      this.model.update(nextProps.annotations, nextProps.viewBoxDimensions);
```

Comment:
`shouldComponentUpdate` should return true/false to indicate whether a new render is needed. It shouldn't be calling object instance methods like this.

Reply:
I think the model update should be called from `componentWillUpdate` or `componentDidUpdate`, leaving `shouldComponentUpdate` to do what it says on the tin, and decide whether or not those methods should be called.

Reply:
The problem is that the model should only update in the case that `shouldComponentUpdate` would return false, and I'm not sure where to do that other than by spawning a Promise.

I've just read in [the react docs](https://reactjs.org/docs/react-component.html#shouldcomponentupdate)

> *"We do not recommend doing deep equality checks or using JSON.stringify() in shouldComponentUpdate(). It is very inefficient and will harm performance."*

So this has to be re-written either way.

Reply:
I'd advise against updating the model when `shouldComponentUpdate` returns false. That's going to confuse anyone who has to debug model updates, and expects updates to run when `shouldComponentUpdate` returns true.